File: js/BogoSort.js
class BogoSort {
  constructor() {
    
  }

  async Sort() {
    while (!this.isSorted() && isStarted){
      await this.shuffle();
    }
  }

  async shuffle() {
    for (let i = 1; i < v.length && isStarted; i++){
      await this.swap(i, parseInt(random(0, 1) * i));
    }
  }

  isSorted() {
    for (let i = 0; i < v.length - 1; i++)
      if (v[i] > v[i - 1])
        return false;
    return true;
  }

  async swap(i, j) {
    status[i] = 2;
    status[j] = 2;

    await sleep();

    let aux = v[i];
    v[i] = v[j];
    v[j] = aux;

    status[i] = 0;
    status[j] = 0;
  }

}

File: src/main/java/window/main/sorting/algorithms/BogoSort.java
package window.main.sorting.algorithms;

import processing.core.PApplet;
import window.config.Complexities;
import window.config.Config;
import window.config.SortingAlgorithms;
import window.main.sorting.SortingAlgorithm;
import window.main.sorting.colors.Color;

public class BogoSort extends SortingAlgorithm {
    public BogoSort(PApplet pApplet, int[] vector, Color[] color) {
        super(pApplet, vector, color, SortingAlgorithms.BOGO_SORT, Complexities.N_N_FACT, Complexities.C);
    }

    @Override
    public void run() {
        onAlgorithmStart();

        bogoSort();

        onAlgorithmStops();
    }

    private void bogoSort() {
        while(!isSorted()) {
            if (stop) {
                return;
            }

            while (pause) {
                if (stop) {
                    return;
                }
                sleep(Config.PAUSE_DELAY_TIME);
            }

            shuffle();

            playSound((int)(Math.random() * 10000) % (vector.length));
            sleep();
            stopSound();
        }
    }

    private void shuffle() {
        int i = vector.length - 1;
        while(i > 0) {
            swapWithNoEffects(i--, (int)(Math.random() * i));
        }
    }

}
